File: MATTHEW/leetcode/AIcoding/4.GPT-4o/java_solutions/2257-count-unguarded-cells-in-the-grid.java
class Solution {
    public int countUnguarded(int m, int n, int[][] guards, int[][] walls) {
        boolean[][] grid = new boolean[m][n];

        for (int[] wall : walls) {
            grid[wall[0]][wall[1]] = true;
        }

        for (int[] guard : guards) {
            grid[guard[0]][guard[1]] = true;
            markUnguarded(grid, guard[0], guard[1], m, n);
        }

        int count = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!grid[i][j]) {
                    count++;
                }
            }
        }

        return count;
    }

    private void markUnguarded(boolean[][] grid, int x, int y, int m, int n) {
        for (int i = x + 1; i < m && !grid[i][y]; i++) {
            grid[i][y] = true;
        }
        for (int i = x - 1; i >= 0 && !grid[i][y]; i--) {
            grid[i][y] = true;
        }
        for (int j = y + 1; j < n && !grid[x][j]; j++) {
            grid[x][j] = true;
        }
        for (int j = y - 1; j >= 0 && !grid[x][j]; j--) {
            grid[x][j] = true;
        }
    }
}

File: MATTHEW/leetcode/AIcoding/5.Qwen-3/rust_solutions/2257-count-unguarded-cells-in-the-grid.rs
struct Solution;

impl Solution {
    pub fn count_unguarded(m: i32, n: i32, guards: Vec<Vec<i32>>, walls: Vec<Vec<i32>>) -> i32 {
        let m = m as usize;
        let n = n as usize;
        let mut grid = vec![vec![0; n]; m];

        for guard in &guards {
            let x = guard[0] as usize;
            let y = guard[1] as usize;
            grid[x][y] = 2; // 2 represents a guard
        }

        for wall in &walls {
            let x = wall[0] as usize;
            let y = wall[1] as usize;
            grid[x][y] = 3; // 3 represents a wall
        }

        let directions = [(0, 1), (1, 0), (0, -1), (-1, 0)];

        for guard in &guards {
            let gx = guard[0] as usize;
            let gy = guard[1] as usize;
            for &(dx, dy) in &directions {
                let mut x = gx as i32 + dx;
                let mut y = gy as i32 + dy;
                while x >= 0 && x < m as i32 && y >= 0 && y < n as i32 {
                    let x_idx = x as usize;
                    let y_idx = y as usize;
                    if grid[x_idx][y_idx] == 0 {
                        grid[x_idx][y_idx] = 1; // 1 represents a cell that is guarded
                    } else if grid[x_idx][y_idx] == 3 {
                        break;
                    }
                    x += dx;
                    y += dy;
                }
            }
        }

        let mut count = 0;
        for i in 0..m {
            for j in 0..n {
                if grid[i][j] == 0 {
                    count += 1;
                }
            }
        }

        count as i32
    }
}